File: _BHCVRP_Python_Version/BHCVRP_Python_Version/cujae_inf_citi_om/generator/postoptimization/Operator_2opt.py
from typing import List
import numpy as np
from postoptimization.StepOptimization import StepOptimization
from solution.Route import Route

class Operator_2opt(StepOptimization):
    def to_optimize(self, route: Route):
        list_opt = route.get_list_id_customers().copy()
        best_cost = route.get_cost_single_route()
        current_cost = 0.0

        for i in range(1, len(list_opt) - 1):
            for j in range(i + 1, len(list_opt)):
                if i < j:  
                    new_route = list_opt[:]
                    new_route[i:j] = reversed(new_route[i:j])
                    
                    new_cost = self.calculate_total_distance(new_route)
                    if new_cost < best_cost:
                        best_cost = new_cost
                        list_opt = new_route

        route.set_list_id_customers(list_opt)
        route.set_cost_route(best_cost)

File: BHCVRP_Java_Version/src/cujae/inf/citi/om/generator/postoptimization/Operator_2opt.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package cujae.inf.citi.om.generator.postoptimization;

import cujae.inf.citi.om.solution.Route;
import java.util.ArrayList;
import java.util.Collections;

/**
 *
 * @author nanda
 */
public class Operator_2opt {
    
    public void toOptimize(Route route) {
        ArrayList<Integer> listOpt = new ArrayList<>(route.getListIdCustomers());
        double bestCost = route.getCostSingleRoute();
        double currentCost = 0.0;

        for (int i = 1; i < listOpt.size() - 1; i++) {
            for (int j = i + 1; j < listOpt.size(); j++) {
                if (i < j) {
                    ArrayList<Integer> newRoute = new ArrayList<>(listOpt);
                    Collections.reverse(newRoute.subList(i, j));

                    double newCost = calculateTotalDistance(newRoute);
                    if (newCost < bestCost) {
                        bestCost = newCost;
                        listOpt = newRoute;
                    }
                }
            }
        }

        route.setListIdCustomers(listOpt);
        route.setCostRoute(bestCost);
    }

    private double calculateTotalDistance(ArrayList<Integer> newRoute) {
        throw new UnsupportedOperationException("Not supported yet."); // Generated from nbfs://nbhost/SystemFileSystem/Templates/Classes/Code/GeneratedMethodBody
    }

    
}
